new_guidance_commands/src/NewGuidanceCommands.cpp

```diff
@@ -0,0 +1,96 @@
+#include "new_guidance_commands/NewGuidanceCommands.hpp"
+
+namespace new_guidance_commands {
+
+NewGuidanceCommands::NewGuidanceCommands(ros::NodeHandle &nodeHandle)
+    : nodeHandle_(nodeHandle)
+{
+  if (!readParameters())
+  {
+    ROS_ERROR("Could not read parameters.");
+    ros::requestShutdown();
+  }
+  speedAccelsubscriber_ = nodeHandle_.subscribe(speedAccel_subscriberTopic_, 1,
+                                      &NewGuidanceCommands::speedAccel_SubscriberCallback, this);
+  wrenchEffortsubscriber_ = nodeHandle_.subscribe(wrenchEffort_subscriberTopic_, 1,
+                                      &NewGuidanceCommands::wrenchEffort_SubscriberCallback, this);
+  lateralControlsubscriber_ = nodeHandle_.subscribe(lateralControl_subscriberTopic_, 1,
+                                      &NewGuidanceCommands::lateralControl_SubscriberCallback, this);
+
+  speedAccel_publisher_ = nodeHandle_.advertise<cav_msgs::SpeedAccel>(speedAccel_publisherTopic_, 1000);
+  wrenchEffort_publisher_ = nodeHandle_.advertise<std_msgs::Float32>(wrenchEffort_publisherTopic_, 1000);
+  lateralControl_publisher_ = nodeHandle_.advertise<cav_msgs::LateralControl>(lateralControl_publisherTopic_, 1000);
+
+  ROS_INFO("Successfully launched node.");
+}
+
+NewGuidanceCommands::~NewGuidanceCommands()
+{
+}
+
+void NewGuidanceCommands::publisher(){
+  ROS_DEBUG("Calling publisher functions");
+  speedAccel_Publisher();
+  wrenchEffort_Publisher();
+  lateralControl_Publisher();
+} 
+
+bool NewGuidanceCommands::readParameters()
+{
+  if (!nodeHandle_.getParam("speedAccel_subscriber_topic", speedAccel_subscriberTopic_) ||
+      !nodeHandle_.getParam("wrenchEffort_subscriber_topic", wrenchEffort_subscriberTopic_) ||
+      !nodeHandle_.getParam("lateralControl_subscriber_topic", lateralControl_subscriberTopic_) ||
+      !nodeHandle_.getParam("speedAccel_publisher_topic", speedAccel_publisherTopic_) ||
+      !nodeHandle_.getParam("wrenchEffort_publisher_topic", wrenchEffort_publisherTopic_) ||
+      !nodeHandle_.getParam("lateralControl_publisher_topic", lateralControl_publisherTopic_) ||
+      !nodeHandle_.getParam("rate", rate)){
+            return false;
+      }
+
+  return true;
+}
+
+
+void NewGuidanceCommands::speedAccel_SubscriberCallback(const cav_msgs::SpeedAccel::ConstPtr& msg){
+    std::lock_guard<std::mutex> lock(SpeedAccel_msg_mutex);
+    SpeedAccel_msg = msg;
+    ROS_DEBUG("I heard SpeedAccel");
+};
+
+void NewGuidanceCommands::wrenchEffort_SubscriberCallback(const std_msgs::Float32::ConstPtr& msg){
+    std::lock_guard<std::mutex> lock(WrenchEffort_msg_mutex);
+    WrenchEffort_msg = msg;
+    ROS_DEBUG("I heard wrenchEffort");
+};
+
+void NewGuidanceCommands::lateralControl_SubscriberCallback(const cav_msgs::LateralControl::ConstPtr& msg){
+    std::lock_guard<std::mutex> lock(LateralControl_msg_mutex);
```

Comment:
From my understanding, we will use mutex lock when there is a multi-threading issue. In our case, we have one thread to write on one variable and the other thread is read-only. Because our write operation here is atomic, I think we can just mark the member variable as volatile. It should be easier than using the lock every time.

Reply:
Hmm, I'm not sure if that solves it. I'm not 100% familiar with the inner workings of the `volatile` keyword, but since copying or accessing these messages isn't by itself an atomic action (it'll take multiple pointer accesses at minimum) I'd think we could still end up reading the first few fields of an old message combined with the last fields of a new one, or similar. Using a mutex like this ensures that the message isn't updated behind the scenes while we're reading it and I'm not sure that volatile works that same way, it just stops the compiler from making assumptions about that variable's state if I understand it correctly.

Reply:
I agreed with what you described. If the write operation is not atomic itself, then using ```volatile``` does not here. But the write operation in this class is accomplished by reassigning a point, I think it should be a single-action operation?